### PR TITLE
Public input circuit optimization

### PR DESCRIPTION
**Pull request by JohnWick2ETH**
There are two kinds of optimization:

1. reduce the length of `rpi` column from about `3*MAX_CALLDATA` to `1*MAX_CALLDATA`.
2. calculate `call_data_gas_cost` using the `value` column in the call_data part of the `tx` table and then use lookup argument to make sure the `call_data_gas_cost` assigned matches the one that is calculated.

### CODE REVIEW

**Comment by lispc** [comment]
the general algorithm optimization seems good. But (1) we'd better make the code style more consistent with the code base. For example: is it possible to use this gadget https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/gadgets/src/is_zero.rs instead of manually construct is_zero expression? (2) usually we update spec before update rust circuit impl

**Comment by CPerezz** [comment]
Happy to review and possibly merge this if you can update the branch @JohnWick2ETH

File: zkevm-circuits/src/pi_circuit.rs
Comment (davidnevadoc) [inline zkevm-circuits/src/pi_circuit.rs:675]: Even though the offsets are computed correctly I think it would be a good idea to set `value_offset` to the start of the `CallDataBytes` portion of the `raw_public_inputs`. So having:
`let value_offset = BLOCK_LEN + 1 + EXTRA_LEN + 3 * tx_table_len;` and reseting `offset=0` when the assignment of CallDataBytes starts.

File: zkevm-circuits/src/pi_circuit.rs
Comment (davidnevadoc) [inline zkevm-circuits/src/pi_circuit.rs:702]: Why is `TxSignHash` removed?

**Review by davidnevadoc — COMMENTED** [review]
Great contirbution! This looks like a good optimization to include.
However, as @lispc mentioned, we should have a the specs modified in zkevm-specs repo first. This way we can check more easily the correctness of the changes and have a broader discussion. 
I suggest we open a PR there to discuss the changes, then we can merge this :)

File: zkevm-circuits/src/pi_circuit.rs
Comment (JohnWick2ETH) [reply zkevm-circuits/src/pi_circuit.rs]: yeah. The comment should be `tx_id_inv`.

File: zkevm-circuits/src/pi_circuit.rs
Comment (JohnWick2ETH) [reply zkevm-circuits/src/pi_circuit.rs]: If the tag is not `CallDataLength` then `tx_id_inv = (tag - CallDataLength)^(-1)`, otherwise, `tx_id_inv` can be any value.

File: zkevm-circuits/src/pi_circuit.rs
Comment (JohnWick2ETH) [reply zkevm-circuits/src/pi_circuit.rs:656]: yeah, it makes sense.

File: zkevm-circuits/src/pi_circuit.rs
Comment (JohnWick2ETH) [reply zkevm-circuits/src/pi_circuit.rs:702]: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/zkevm-circuits/src/witness/tx.rs#L41-L118

**Comment by JohnWick2ETH** [comment]
> 
I've opened an pr in zkevm-specs repo.  privacy-scaling-explorations/zkevm-specs#277

**Review by davidnevadoc — APPROVED** [review]
Just a couple of leftover comments, all the rest LGTM!
Great job! :)

File: zkevm-circuits/src/pi_circuit.rs
Comment (kunxian-xia) [reply zkevm-circuits/src/pi_circuit.rs]: fixed.

File: zkevm-circuits/src/pi_circuit.rs
Comment (kunxian-xia) [reply zkevm-circuits/src/pi_circuit.rs]: fixed.

File: zkevm-circuits/src/pi_circuit.rs
Comment (lispc) [inline zkevm-circuits/src/pi_circuit.rs:1487]: a small question: why increase these two values and k?   
(bigger k needs longer CI time, so usually we tends to use smaller cases unless needed otherwise)

@kunxian-xia

Comment (lispc) [reply zkevm-circuits/src/pi_circuit.rs:1487]: oh it is due to u16 lookup table

**Comment by lispc** [comment]
a suggestion: add several lines to explain the u16 fixed lookup table: to ensure the monotonicity of tx_id. (u8 is not safe enough 256 * 21000(transfer gas) is less than block gas limit).

**Review by lispc — APPROVED** [review]

**Comment by CPerezz** [comment]
This got all the approvals required.
Please @JohnWick2ETH could you resolve the conflicts so that we can merge this???

This PR is currently blocking the merge of #935

**Pull request merged**
